crates/ruff_python_parser/src/lexer.rs

```diff
@@ -1491,19 +1489,39 @@ const fn is_ascii_identifier_start(c: char) -> bool {
 
 // Checks if the character c is a valid starting character as described
 // in https://docs.python.org/3/reference/lexical_analysis.html#identifiers
+#[cold]
 fn is_unicode_identifier_start(c: char) -> bool {
-    is_xid_start(c)
+    unicode_ident::is_xid_start(c)
 }
 
 // Checks if the character c is a valid continuation character as described
 // in https://docs.python.org/3/reference/lexical_analysis.html#identifiers
 fn is_identifier_continuation(c: char) -> bool {
-    match c {
-        'a'..='z' | 'A'..='Z' | '_' | '0'..='9' => true,
-        c => is_xid_continue(c),
+    if c.is_ascii() {
+        ASCII_CONTINUE.0[c as usize]
+    } else {
+        is_unicode_identifier_continue(c)
```

Comment:
If you're looking to speed things up here for the ASCII fast path, it might make sense to refactor the calling code to deal with this `&str` chunks instead. That way, you can get entire identifiers all in one go if that's the common path. (But if it's not the overwhelmingly common path, and it might not be, then it may not be such a good idea.) This _could_ also be a good "started SIMD" task.

Reply:
This sounds interesting even just to learn. Can you say a bit more here? What's the chunk?

Reply:
To be clear, I don't have any understanding of what the caller code looks like. So the idea may be bunk. But basically, if you're calling this function over and over on each individual codepoint, then you're paying a price for the latency of doing that check. If it's a _necessary_ payment, then there is not much to do. But if "a starting identifier char followed by continuation chars" is extremely common, then it may make more sense to have an API like this:

```rust
fn get_ident_from_prefix(src: &str) -> Option<Identifier> { ... }
```

The idea is that if you're _pretty_ sure there's an identifier there in most cases, then you make the uncommon cases a little slower in favor of making the common case a lot faster. The idea here is that you might be able to detect identifiers very quickly with bit tricks and/or SIMD. The specific details of what the bit tricks/SIMD are is left as an exercise to for the reader. :)

Reply:
The calling code is very simple. We call it for every character until the end of an identifier.



```
        self.cursor.eat_while(is_identifier_continuation);

```


It eats the characters why `is_identifier_continuation` is true, and we expect most characters to be ascii. There are probably some SIMD tricks that we could apply but it never felt worthwhile considering that lexing used to be a very small fraction compared to parsing.


It could make sense to add an explicit `c if c.is_ascii() => false` case to avoid calling into the function when we reach any space character (the end of an identifier).

Reply:
Oh yeah absolutely, if lexing is already a small slice of the pie, it may indeed not make sense to expend a lot of effort here.

Reply:
It might be worth it with the new parser ;)